### PR TITLE
[FLINK-26618][sql-client] Fix Remove-Jar statement not effective

### PR DESCRIPTION
**Pull request by link3280**
## What is the purpose of the change

Currently, `remove jar` statement doesn't remove the corresponding jars in pipeline.jars, which should be fixed.

## Brief change log

- Replacing session dependencies instead of merging them into the new dependencies in remove-jar statements.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by link3280** [comment]
CI failures are irrelevant, retrigger the pipeline.

**Comment by link3280** [comment]
@flinkbot run azure

**Comment by link3280** [comment]
@flinkbot run azure

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java [hunk 1]
```diff
@@ -261,8 +261,20 @@ public void addJar(String jarPath) {
...
+        Set<URL> jarsInConfig;
+        try {
+            jarsInConfig =
+                    new HashSet<>(
+                            ConfigUtils.decodeListFromConfig(
+                                    sessionConfiguration, PipelineOptions.JARS, URL::new));
+        } catch (MalformedURLException e) {
+            throw new SqlExecutionException(
+                    "Failed to parse the option `pipeline.jars` in configuration.", e);
+        }
```
Comment (fsk119) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java:274]: I think the semantic of the ADD JAR syntax does two things:
1. ADD the jar URL into the classloader dependencies;
2. ADD the URL into the config `pipeline.jars` for later upload;

Therefore, I think the REMOVE JAR also needs to do as the ADD JAR does.
1. REMOVE the jar URL from the classloader dependencies;
2. REMOVE the URL in the config `pipeline.jars` ;

The main problem is that user may try to SET pipeline.jars in SQL Client. At this point, I think we can't just overwrite the config with the maintained dependencies in the classloader when REMOVE JAR, which causes the user's setting is different with before. I think we should also merge the config and then try to remove. What do you think?

**Review by fsk119 — COMMENTED** [review]
Thanks for your great work. 

Could you also add a case in the CliClientITCase? You can take a look at the file set.q[1] for more details. 

[1] https://github.com/apache/flink/blob/master/flink-table/flink-sql-client/src/test/resources/sql/set.q#L128

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java (hunk 1 above)
Comment (link3280) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java:274]: @fsk119 Thanks a lot for your pointers! It clears lots of my doubts. Updating the PR.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java [hunk 2]
```diff
@@ -261,7 +261,20 @@ public void addJar(String jarPath) {
...
```
Comment (fsk119) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/context/SessionContext.java:274]: nit: Would be better to move these codes to the single helper method.

**Review by fsk119 — COMMENTED** [review]
Thanks for your update. Leave one minor suggestion.

**Comment by fsk119** [comment]
cc @leonardBang @wuchong

**Review by fsk119 — APPROVED** [review]
Thanks for your update. LGTM

**Comment by link3280** [comment]
Thanks a lot for your review! @fsk119

**Review by leonardBang — APPROVED** [review]
+1

**Comment by leonardBang** [comment]
@flinkbot run azure

**Comment by leonardBang** [comment]
@link3280 Could you also open PRs for branch `release-1.14` and `release-1.15` ?

**Comment by link3280** [comment]
@leonardBang Sure.

**Comment by leonardBang** [comment]
@flinkbot run azure

**Comment by link3280** [comment]
@leonardBang Opened 2 backport PR #19132 #19133

**Comment by leonardBang** [comment]
> @leonardBang Opened 2 backport PR #19132 #19133

Nice work！We need rebase this PR to latest master to pass the e2e tests, the e2e tests failed today and has been fixed in
https://github.com/apache/flink/commit/cbf2ab0ea0f481df9e85b2531bf07e1cd5c8ca91

**Comment by link3280** [comment]
@leonardBang Done. Updating the other two PRs.

**Comment by link3280** [comment]
@flinkbot run azure

**Comment by link3280** [comment]
@flinkbot run azure

**Comment by link3280** [comment]
@leonardBang CI finally turns green. Could you merge this PR if possible? Thanks.

**Comment by leonardBang** [comment]
Thanks @link3280 for the continuous work, the Github infrastructure broken recently which influenced many project 😂.

**Pull request merged**